pkg/resto/resto.go

```diff
@@ -0,0 +1,357 @@
+	if err != nil {
+		return "", "", err
+	}
+	err = configFile.LoadFromReader(cr)
+	if err != nil {
+		return "", "", err
+	}
+	cr.Close()
+	if domain == "https://registry-1.docker.io" {
+		domain = "https://index.docker.io/v1/"
+	}
+	domain = strings.TrimPrefix(domain, "https://")
+	auth, err := configFile.GetAuthConfig(domain)
+	if err != nil {
+		//fmt.Printf("GetAuthConfig failure for %s: %s\n", domain, err)
+		return "", "", err
+	}
+	return auth.Username, auth.Password, nil
+}
+
+func makeTarGz(content map[string]string) ([]byte, digest.Digest, error) {
+	buf := bytes.NewBuffer(nil)
+	w := tar.NewWriter(buf)
+	for k, v := range content {
+		if err := w.WriteHeader(&tar.Header{
+			Typeflag: tar.TypeReg,
+			Name:     k,
+			Mode:     0600,
+			Size:     int64(len(v)),
+		}); err != nil {
+			return nil, "", err
+		}
+		if _, err := w.Write([]byte(v)); err != nil {
+			return nil, "", err
+		}
+	}
+	w.Close()
+	dgst := digest.SHA256.FromBytes(buf.Bytes())
+	gzbuf := bytes.NewBuffer(nil)
+	g := gzip.NewWriter(gzbuf)
+	g.Write(buf.Bytes())
+	g.Close()
+	return gzbuf.Bytes(), dgst, nil
+}
+
+// ListRegistry lists all the repositories in a registry
+func ListRegistry(ctx context.Context, endpoint, username, password string, insecure bool) ([]string, error) {
+	tr, err := NewTransportCatalog(endpoint, username, password, insecure)
+	if err != nil {
+		return nil, err
+	}
+	registry, err := client.NewRegistry(endpoint, tr)
+	if err != nil {
+		return nil, err
+	}
+	entries := make([]string, 10000)
+	count, err := registry.Repositories(ctx, entries, "")
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	return entries[0:count], nil
+}
+
+// ListRepository lists all the tags in a repository
+func ListRepository(ctx context.Context, reponame, username, password string, insecure bool) ([]string, error) {
+	pr, err := parseRef(reponame)
+	if err != nil {
+		return nil, err
+	}
+	repo, err := NewRepository(ctx, pr.domain, pr.path, username, password, insecure)
+	if err != nil {
+		return nil, err
+	}
+	tagService := repo.Tags(ctx)
+	return tagService.All(ctx)
+}
+
+// PullConfig pulls a configuration file from a registry
+func PullConfig(ctx context.Context, repoTag, username, password string, insecure bool) (string, error) {
+	res, err := PullConfigMulti(ctx, repoTag, username, password, insecure)
+	if err != nil {
+		return "", err
+	}
+	return res["config"], nil
+}
+
+// PullConfigMulti pulls a set of configuration files from a registry
+func PullConfigMulti(ctx context.Context, repoTag, username, password string, insecure bool) (map[string]string, error) {
+	pr, err := parseRef(repoTag)
+	if err != nil {
+		return nil, err
+	}
+	if username == "" {
+		username, password, _ = getCredentials(pr.domain)
+	}
+	repo, err := NewRepository(ctx, pr.domain, pr.path, username, password, insecure)
+	if err != nil {
+		repo, err = NewRepository(ctx, strings.Replace(pr.domain, "https://", "http://", 1), pr.path, username, password, insecure)
```

Comment:
This looks fishy; something failed (we don't know what), so leak credentials over  plain `http://`?

Reply:
I changed it so that it drops credentials by default when fallbacking to http, and so that it fallbacks only if the error suggests we should.